### PR TITLE
ATT-43: AttachmentResource to accept Base64 POST content.

### PR DESCRIPTION
**Pull request by ivange94**
https://issues.openmrs.org/browse/ATT-43

### CODE REVIEW

File: omod-1.10/pom.xml [hunk 1]
```diff
@@ -184,6 +184,16 @@
         <filtering>false</filtering>
       </testResource>
     </testResources>
+      <plugins>
+          <plugin>
+              <groupId>org.apache.maven.plugins</groupId>
+              <artifactId>maven-compiler-plugin</artifactId>
+              <configuration>
+                  <source>8</source>
+                  <target>8</target>
```
Comment (mks-d) [inline omod-1.10/pom.xml]: I don't think it's ok to enforce Java 8 against Core 1.10.

File: omod-1.10/src/main/java/org/openmrs/module/attachments/rest/AttachmentResource1_10.java [hunk 2]
```diff
@@ -102,7 +109,11 @@ public Object upload(MultipartFile file, RequestContext context) throws Response
 		Provider provider = Context.getProviderService().getProviderByUuid(context.getParameter("provider"));
 		String fileCaption = context.getParameter("fileCaption");
 		String instructions = context.getParameter("instructions");
+		String image = context.getParameter("cameraImage");
```
Comment (mks-d) [inline omod-1.10/src/main/java/org/openmrs/module/attachments/rest/AttachmentResource1_10.java]: Let's do:
```java
String base64Content = context.getParameter("base64Content");
```
The reason being that Base64 can be used to encode pretty much anything, not just images.

**Comment by HerbertYiga** [comment]
hi @ivange94   well done on this,could we go on to create at a ticket on this

File: omod-1.10/src/main/java/org/openmrs/module/attachments/rest/AttachmentResource1_10.java [hunk 3]
```diff
@@ -255,4 +266,81 @@ protected PageableResult doSearch(RequestContext context) {
...
+	 * Wrapper class to be passed to ComplexObsSaver#saveImageAttachment
+	 * ComplexObsSaver#saveImageAttachment needs a MultipartFile but the image could either be a
+	 * MultipartFile or a base64 encoded image. This class will only implement the methods used by
+	 * ComplexObsSaver#saveImageAttachment. This way we won't have to make any changes to the
+	 * implementation of ComplexObsSaver#saveImageAttachment and will also make very little change to
+	 * the AttachmentResource1_10#upload implementation. This is also helps us avoid adding an extra
+	 * dependency to MockMultipartFile for converting the base64 encoded String to a MultipartFile
+	 * object.
+	 */
+	class MultipartFileWrapper implements MultipartFile {
```
Comment (mks-d) [inline omod-1.10/src/main/java/org/openmrs/module/attachments/rest/AttachmentResource1_10.java]: Name it `Base64MultipartFile`.

**Review by mks-d — COMMENTED** [review]
A couple of comments.

We will also need unit test(s) that show(s) that all is well with this new part of the implementation.

File: omod-1.10/src/test/java/org/openmrs/module/attachments/rest/AttachmentRestController1_10Test.java [hunk 4]
```diff
@@ -281,6 +285,57 @@ public void postAttachment_shouldUploadFileToEncounter() throws Exception {
...
+		request.addParameter("patient", patient.getUuid());
+		request.addParameter("fileCaption", fileCaption);
+		request.addParameter("base64Content", base64Content);
+		
+		// Replay
+		SimpleObject response = deserialize(handle(request));
+		
+		Obs obs = Context.getObsService().getObsByUuid((String) response.get("uuid"));
+		Obs complexObs = Context.getObsService().getComplexObs(obs.getObsId(), null);
+		ComplexData complexData = complexObs.getComplexData();
```
Comment (mks-d) [inline omod-1.10/src/test/java/org/openmrs/module/attachments/rest/AttachmentRestController1_10Test.java:323]: Now that you've got the complex data of the saved obs, you can get the bytes right away with [this](https://github.com/openmrs/openmrs-module-attachments/blob/91a848d462c989b0d3669f63d4a37eb2df9c6cff/api/src/main/java/org/openmrs/module/attachments/obs/BaseComplexData.java#L36) util method. And then you could encode this as Base64 and assert that you got back the same as `base64Content`.

File: omod-1.10/src/test/java/org/openmrs/module/attachments/rest/AttachmentRestController1_10Test.java [hunk 5]
```diff
@@ -281,6 +285,57 @@ public void postAttachment_shouldUploadFileToEncounter() throws Exception {
 		Assert.assertEquals(obs.getEncounter().getUuid(), encounter.getUuid());
 	}
 	
+	@Test
+	public void postAttachment_shouldAcceptBase64Content() throws Exception {
+		String fileCaption = "Test file caption";
+		String fileName = "testFile2.dat";
```
Comment (mks-d) [inline omod-1.10/src/test/java/org/openmrs/module/attachments/rest/AttachmentRestController1_10Test.java:306]: Maybe `"test-image.png"`?

File: omod-1.10/src/test/java/org/openmrs/module/attachments/rest/AttachmentRestController1_10Test.java [hunk 6]
```diff
@@ -281,6 +285,57 @@ public void postAttachment_shouldUploadFileToEncounter() throws Exception {
...
+		
+		// Verify
+		Assert.assertEquals(obs.getComment(), fileCaption);
+		Assert.assertTrue(complexData.getTitle().startsWith("cameracapture"));
+		Assert.assertTrue(compareImages(imgIn, imgOut));
+		Assert.assertNull(obs.getEncounter());
+		
+	}
+	
+	private boolean compareImages(BufferedImage imgIn, BufferedImage imgOut) {
```
Comment (mks-d) [inline omod-1.10/src/test/java/org/openmrs/module/attachments/rest/AttachmentRestController1_10Test.java]: So this method shouldn't be necessary anymore.

File: omod-1.10/src/test/java/org/openmrs/module/attachments/rest/AttachmentRestController1_10Test.java (hunk 4 above)
Comment (ivange94) [reply omod-1.10/src/test/java/org/openmrs/module/attachments/rest/AttachmentRestController1_10Test.java:323]: For some reason the encodings are not the same. When I used bufferedimage, the pixels matched but when I try to encode the `byte[]` to a base64 string, they're not equivalent. 

Screen Shot 2020-06-22 at 14 19 09

Comment (ivange94) [reply omod-1.10/src/test/java/org/openmrs/module/attachments/rest/AttachmentRestController1_10Test.java:308]: The bytes returned by `BaseComplexData.getByteArray(complexData);` doesn't match that returned by `Base64.decodeBase64(base64Conten);`

Comment (ivange94) [reply omod-1.10/src/test/java/org/openmrs/module/attachments/rest/AttachmentRestController1_10Test.java:323]: The bytes returned by `BaseComplexData.getByteArray(complexData);` does not match that returned by `Base64.decodeBase64(base64Content.split(",")[1].trim().getBytes());`

**Review by mks-d — APPROVED** [review]

**Pull request merged**